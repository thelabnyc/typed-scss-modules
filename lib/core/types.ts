import type { SASSOptions } from "../sass/index.ts";
import type { ExportType, LogLevel, QuoteType } from "../typescript/index.ts";

type ConfigFileOnlyOptions = Extract<keyof SASSOptions, "importers">;

export interface CLIOptions
    extends Exclude<SASSOptions, ConfigFileOnlyOptions> {
    banner: string;
    ignore: string[];
    ignoreInitial: boolean;
    exportType: ExportType;
    exportTypeName: string;
    exportTypeInterface: string;
    listDifferent: boolean;
    quoteType: QuoteType;
    updateStaleOnly: boolean;
    watch: boolean;
    logLevel: LogLevel;
    outputFolder: string | null;
    allowArbitraryExtensions: boolean;
}

export interface ConfigOptions extends CLIOptions, SASSOptions {}
