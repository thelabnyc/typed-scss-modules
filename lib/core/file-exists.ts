import fs from "node:fs/promises";

import { constants } from "fs";

export const fileExists = async (path: string): Promise<boolean> => {
    try {
        await fs.access(path, constants.F_OK);
        return true;
    } catch {
        return false;
    }
};
