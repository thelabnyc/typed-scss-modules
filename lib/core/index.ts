export { alerts, setAlertsLogLevel } from "./alerts.ts";
export { generate } from "./generate.ts";
export { listDifferent } from "./list-different.ts";
export type { CLIOptions, ConfigOptions } from "./types.ts";
export { watch } from "./watch.ts";
export { writeFile } from "./write-file.ts";
