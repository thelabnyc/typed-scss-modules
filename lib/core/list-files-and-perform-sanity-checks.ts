import { glob } from "glob";

import { alerts } from "./alerts.ts";
import type { ConfigOptions } from "./types.ts";

/**
 * Return the files matching the given pattern and alert the user if only 0 or 1
 * files matched.
 *
 * @param pattern the file pattern to generate type definitions for
 * @param options the CLI options
 */
export const listFilesAndPerformSanityChecks = async (
    pattern: string,
    options: ConfigOptions,
): Promise<string[]> => {
    // Find all the files that match the provided pattern.
    const files = await glob(pattern, { ignore: options.ignore });
    if (!files.length) {
        alerts.error("No files found.");
    }
    // This case still works as expected but it's easy to do on accident so
    // provide a (hopefully) helpful warning.
    if (files.length === 1) {
        alerts.warn(
            `Only 1 file found for ${pattern}. If using a glob pattern (eg: dir/**/*.scss) make sure to wrap in quotes (eg: "dir/**/*.scss").`,
        );
    }
    return files;
};
