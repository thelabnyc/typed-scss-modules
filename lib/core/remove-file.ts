import fs from "node:fs";

import { getTypeDefinitionPath } from "../typescript/index.ts";
import { alerts } from "./alerts.ts";
import type { ConfigOptions } from "./index.ts";

/**
 * Given a single file remove the file
 *
 * @param file any file to remove
 */

const removeFile = (file: string): void => {
    try {
        if (fs.existsSync(file)) {
            fs.unlinkSync(file);
            alerts.success(`[REMOVED] ${file}`);
        }
    } catch (error) {
        alerts.error(
            `An error occurred removing ${file}:\n${JSON.stringify(error)}`,
        );
    }
};

/**
 * Given a single file remove the generated types if they exist
 *
 * @param file the SCSS file to generate types for
 */
export const removeSCSSTypeDefinitionFile = (
    file: string,
    options: ConfigOptions,
): void => {
    const path = getTypeDefinitionPath(file, options);
    removeFile(path);
};
