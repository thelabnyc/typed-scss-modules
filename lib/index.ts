export { main as default } from "./main.ts";
