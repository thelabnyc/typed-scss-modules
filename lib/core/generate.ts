import { alerts } from "./alerts.ts";
import { listFilesAndPerformSanityChecks } from "./list-files-and-perform-sanity-checks.ts";
import type { ConfigOptions } from "./types.ts";
import { writeFile } from "./write-file.ts";

/**
 * Given a file glob generate the corresponding types once.
 *
 * @param pattern the file pattern to generate type definitions for
 * @param options the CLI options
 */
export const generate = async (
  pattern: string,
  options: ConfigOptions
): Promise<void> => {
  const files = listFilesAndPerformSanityChecks(pattern, options);

  if (files.length === 0) {
    return;
  }

  alerts.success(
    `Found ${files.length} file${
      files.length === 1 ? `` : `s`
    }. Generating type definitions...`
  );

  // Wait for all the type definitions to be written.
  await Promise.all(files.map((file) => writeFile(file, options)));
};
