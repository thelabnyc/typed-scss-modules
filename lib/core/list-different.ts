import fs from "node:fs/promises";

import { fileToClassNames } from "../sass/index.ts";
import {
    classNamesToTypeDefinitions,
    getTypeDefinitionPath,
} from "../typescript/index.ts";
import { alerts } from "./alerts.ts";
import { fileExists } from "./file-exists.ts";
import { listFilesAndPerformSanityChecks } from "./list-files-and-perform-sanity-checks.ts";
import type { ConfigOptions } from "./types.ts";

export const checkFile = async (
    file: string,
    options: ConfigOptions,
): Promise<boolean> => {
    try {
        const classNames = await fileToClassNames(file, options);
        const typeDefinition = await classNamesToTypeDefinitions({
            classNames: classNames,
            file,
            ...options,
        });

        if (!typeDefinition) {
            // Assume if no type defs are necessary it's fine
            return true;
        }

        const path = getTypeDefinitionPath(file, options);
        if (!(await fileExists(path))) {
            alerts.error(
                `[INVALID TYPES] Type file needs to be generated for ${file} `,
            );
            return false;
        }

        const content = await fs.readFile(path, { encoding: "utf8" });
        if (content !== typeDefinition) {
            alerts.error(`[INVALID TYPES] Check type definitions for ${file}`);
            return false;
        }

        return true;
    } catch (error) {
        alerts.error(
            `An error occurred checking ${file}:\n${(error as Error).toString()}`,
        );
        return false;
    }
};

export const listDifferent = async (
    pattern: string,
    options: ConfigOptions,
): Promise<void> => {
    const files = await listFilesAndPerformSanityChecks(pattern, options);

    // Wait for all the files to be checked.
    const validChecks = await Promise.all(
        files.map((file) => checkFile(file, options)),
    );
    if (validChecks.includes(false)) {
        process.exit(1);
    }
};
