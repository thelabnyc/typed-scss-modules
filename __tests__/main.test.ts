import fs from "fs";
import path from "path";
import slash from "slash";
import { alerts } from "../lib/core";
import { main } from "../lib/main";

describe("main", () => {
  let writeFileSyncSpy: jest.SpyInstance;

  beforeEach(() => {
    // Only mock the writes, so the example files can still be read.
    writeFileSyncSpy = jest.spyOn(fs, "writeFileSync").mockImplementation();

    // Avoid creating directories while running tests.
    jest.spyOn(fs, "mkdirSync").mockImplementation();

    // Avoid console logs showing up.
    jest.spyOn(console, "log").mockImplementation();

    jest.spyOn(alerts, "error").mockImplementation();
  });

  afterEach(() => {
    writeFileSyncSpy.mockReset();
  });

  it("generates types for all .scss files when the pattern is a directory", async () => {
    const pattern = `${__dirname}/dummy-styles`;

    await main(pattern, {
      banner: "",
      watch: false,
      ignoreInitial: false,
      exportType: "named",
      exportTypeName: "ClassNames",
      exportTypeInterface: "Styles",
      listDifferent: false,
      ignore: [],
      implementation: "sass",
      quoteType: "single",
      updateStaleOnly: false,
      logLevel: "verbose",
      additionalData: "$global-red: red;",
      aliases: {
        "~fancy-import": "complex",
        "~another": "style",
      },
      aliasPrefixes: {
        "~": "nested-styles/",
      },
    });

    expect(alerts.error).not.toHaveBeenCalled();
    expect(fs.writeFileSync).toHaveBeenCalledTimes(9);

    const expectedDirname = slash(path.join(__dirname, "dummy-styles"));

    expect(fs.writeFileSync).toHaveBeenCalledWith(
      `${expectedDirname}/complex.scss.d.ts`,
      "export declare const nestedAnother: string;\nexport declare const nestedClass: string;\nexport declare const number1: string;\nexport declare const someStyles: string;\nexport declare const whereSelector: string;\n"
    );
    expect(fs.writeFileSync).toHaveBeenCalledWith(
      `${expectedDirname}/style.scss.d.ts`,
      "export declare const someClass: string;\n"
    );
  });

  it("generates types for all .scss files and ignores files that match the ignore pattern", async () => {
    const pattern = `${__dirname}/dummy-styles`;

    await main(pattern, {
      banner: "",
      watch: false,
      ignoreInitial: false,
      exportType: "named",
      exportTypeName: "ClassNames",
      exportTypeInterface: "Styles",
      listDifferent: false,
      ignore: ["**/style.scss"],
      implementation: "sass",
      quoteType: "single",
      updateStaleOnly: false,
      logLevel: "verbose",
      additionalData: "$global-red: red;",
      aliases: {
        "~fancy-import": "complex",
        "~another": "style",
      },
      aliasPrefixes: {
        "~": "nested-styles/",
      },
    });

    expect(alerts.error).not.toHaveBeenCalled();
    expect(fs.writeFileSync).toHaveBeenCalledTimes(7);

    const expectedDirname = slash(path.join(__dirname, "dummy-styles"));

    expect(fs.writeFileSync).toHaveBeenCalledWith(
      `${expectedDirname}/complex.scss.d.ts`,
      "export declare const nestedAnother: string;\nexport declare const nestedClass: string;\nexport declare const number1: string;\nexport declare const someStyles: string;\nexport declare const whereSelector: string;\n"
    );

    // Files that should match the ignore pattern.
    expect(fs.writeFileSync).not.toHaveBeenCalledWith(
      `${expectedDirname}/style.scss.d.ts`,
      expect.anything()
    );
    expect(fs.writeFileSync).not.toHaveBeenCalledWith(
      `${expectedDirname}/nested-styles/style.scss.d.ts`,
      expect.anything()
    );
  });

  it("reads options from the configuration file", async () => {
    const pattern = `${__dirname}/dummy-styles`;

    jest.spyOn(process, "cwd").mockReturnValue(path.resolve(pattern));

    await main(pattern, {
      additionalData: "$global-red: red;",
      aliases: {
        "~fancy-import": "complex",
        "~another": "style",
      },
      aliasPrefixes: {
        "~": "nested-styles/",
      },
      exportType: "default",
    });

    expect(alerts.error).not.toHaveBeenCalled();
    expect(fs.writeFileSync).toHaveBeenCalledTimes(9);

    // Transform the calls into a more readable format for the snapshot.
    const contents = writeFileSyncSpy.mock.calls
      .map(([fullFilePath, contents]: [string, string]) => ({
        path: path.relative(__dirname, fullFilePath),
        contents,
      }))
      // Sort to avoid flakey snapshot tests if call order changes.
      .sort((a, b) => a.path.localeCompare(b.path));

    expect(contents).toMatchSnapshot();
  });

  it("outputs the correct files when outputFolder is passed", async () => {
    const pattern = path.resolve(__dirname, "dummy-styles");

    await main(pattern, {
      additionalData: "$global-red: red;",
      aliases: {
        "~fancy-import": "complex",
        "~another": "style",
      },
      aliasPrefixes: {
        "~": "nested-styles/",
      },
      outputFolder: "__generated__",
    });

    expect(alerts.error).not.toHaveBeenCalled();
    expect(fs.writeFileSync).toHaveBeenCalledTimes(9);
    expect(fs.mkdirSync).toHaveBeenCalledTimes(9);

    // Transform the calls into a more readable format for the snapshot.
    const contents = writeFileSyncSpy.mock.calls
      .map(([fullFilePath, contents]: [string, string]) => ({
        path: path.relative(__dirname, fullFilePath),
        contents,
      }))
      // Sort to avoid flakey snapshot tests if call order changes.
      .sort((a, b) => a.path.localeCompare(b.path));

    expect(contents).toMatchSnapshot();
  });
});
