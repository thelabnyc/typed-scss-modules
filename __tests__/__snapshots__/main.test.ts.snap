// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main outputs the correct files when outputFolder is passed 1`] = `
[
  {
    "contents": "export declare const myCustomClass: string;
export declare const nestedAnother: string;
export declare const nestedClass: string;
export declare const nestedStyles: string;
export declare const number1: string;
export declare const someStyles: string;
export declare const whereSelector: string;
",
    "path": "../__generated__/__tests__/dummy-styles/alias-prefixes.scss.d.ts",
  },
  {
    "contents": "export declare const myCustomClass: string;
export declare const nestedAnother: string;
export declare const nestedClass: string;
export declare const number1: string;
export declare const someClass: string;
export declare const someStyles: string;
export declare const whereSelector: string;
",
    "path": "../__generated__/__tests__/dummy-styles/aliases.scss.d.ts",
  },
  {
    "contents": "export declare const nestedAnother: string;
export declare const nestedClass: string;
export declare const number1: string;
export declare const someStyles: string;
export declare const whereSelector: string;
",
    "path": "../__generated__/__tests__/dummy-styles/complex.scss.d.ts",
  },
  {
    "contents": "export declare const composedClass: string;
",
    "path": "../__generated__/__tests__/dummy-styles/composes.scss.d.ts",
  },
  {
    "contents": "export declare const app: string;
export declare const appHeader: string;
export declare const logo: string;
",
    "path": "../__generated__/__tests__/dummy-styles/dashes.scss.d.ts",
  },
  {
    "contents": "export declare const globalStyle: string;
",
    "path": "../__generated__/__tests__/dummy-styles/global-variables.scss.d.ts",
  },
  {
    "contents": "export declare const randomClass: string;
",
    "path": "../__generated__/__tests__/dummy-styles/invalid.scss.d.ts",
  },
  {
    "contents": "export declare const nestedStyles: string;
",
    "path": "../__generated__/__tests__/dummy-styles/nested-styles/style.scss.d.ts",
  },
  {
    "contents": "export declare const someClass: string;
",
    "path": "../__generated__/__tests__/dummy-styles/style.scss.d.ts",
  },
]
`;

exports[`main reads options from the configuration file 1`] = `
[
  {
    "contents": "export type Styles = {
  myCustomClass: string;
  nestedAnother: string;
  nestedClass: string;
  nestedStyles: string;
  number1: string;
  someStyles: string;
  whereSelector: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/alias-prefixes.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  myCustomClass: string;
  nestedAnother: string;
  nestedClass: string;
  number1: string;
  someClass: string;
  someStyles: string;
  whereSelector: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/aliases.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  nestedAnother: string;
  nestedClass: string;
  number1: string;
  someStyles: string;
  whereSelector: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/complex.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  composedClass: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/composes.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  app: string;
  appHeader: string;
  logo: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/dashes.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  globalStyle: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/global-variables.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  randomClass: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/invalid.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  nestedStyles: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/nested-styles/style.scss.d.ts",
  },
  {
    "contents": "export type Styles = {
  someClass: string;
};

export type ClassNames = keyof Styles;

declare const styles: Styles;

export default styles;
",
    "path": "dummy-styles/style.scss.d.ts",
  },
]
`;
